### ISSUE TITLE
Serialization of `PermissibleValue` is wrong

### ISSUE THREAD

**Issue by Silvanoc** [linkml/linkml#2382]
**Describe the bug**

The serialization of a `PermissibleValue` like `PermissibleValue('infant')` should be `'infant'`, instead it is `PermissibleValue({'text': 'infant'})`.

This issue is probably the root-cause of issue #2380 

**Version of LinkML you are using** v1.8.5

**Please provide a schema (and if applicable, a data file) that replicates the issue**
No schema, but some code snippet to reproduce it:

``` python
#!/usr/bin/env python3

from linkml_runtime.linkml_model.meta import PermissibleValue

value = PermissibleValue('infant')
print(value)
```

It prints `PermissibleValue({'text': 'infant'})` and not `'infant'` (what I would expect).

**Comment by Silvanoc** [comment]
@sneakers-the-rat you have been active lately in that part of the code. Could you please confirm the issue?

**Comment by sneakers-the-rat** [comment]
aha, yeah well when i wrote that I didn't know that we were using those models for literals like enum values. should be simple enough to fix by just removing the `__str__` method here: https://github.com/linkml/linkml-runtime/blob/ad74966db72be214a38de19f361df88489286e83/linkml_runtime/utils/yamlutils.py#L286

**Comment by sneakers-the-rat** [comment]
> It is unclear why no tests (neither in linkml nor in linkml-runtime) have raised this issue

this is the bigger problem, i don't know much about the json schema generator but i don't think that any generator should be doing blind type coercions like `str(NonstandardType)`. Having the `__str__` method be overgeneral is definitely my bad, but i'm concerned that the compliance tests didn't catch this. it seems like this test should have?

 https://github.com/linkml/linkml/blob/08959cb73d8b09d66bbad730bb5f4d63f0195b08/tests/test_compliance/test_enum_compliance.py#L44
 
actually now that i look i'm not sure what that's happening actually, because i don't see a `str` call in the generator, i see a https://github.com/linkml/linkml/blob/08959cb73d8b09d66bbad730bb5f4d63f0195b08/linkml/generators/jsonschemagen.py#L383

which should handle the value correctly
 
 So i'm gonna return in the morning because I was about to go to bed - it is true that the value of an enum that's expected to behave as a string like `PermissibleValue` should behave like `str(PermissibleValue('someValue')) == 'someValue'` , in general I think that type-like values should absolutely have their own generation logic so they behave like the types they emulate - `enum` values should be immutable values, ideally that can be unsurprisingly coerced to the represented value.
 
 I am not sure that the fact that the fact that `str(<Element>)` doesn't produce the expected value is the cause of [this bug](https://github.com/linkml/linkml/actions/runs/11513226137/job/32049565431?pr=2381#step:9:365) from [this change](https://github.com/linkml/linkml-model/commit/e68c4f3c2094e2b6b3788384e21ec00ce1b9c3e2) though, because that looks totally unrelated, but i'll take a look tmrw

**Comment by sneakers-the-rat** [comment]
problem is that the `meta.py` file is generated by the python generator, so that PR would be overwritten the next time the metamodel was generated. 

If that is in fact the root of the problem you're seeing, we would need to adjust the python generator logic (and ideally the root enum types and the compliance tests).

**Comment by sneakers-the-rat** [comment]
like even just looking at the CI run error it's clear that it's more than a `__str__` error: https://github.com/linkml/linkml/actions/runs/11513226137/job/32049565431?pr=2381#step:9:387

```
    def test_infer_and_convert(input_path, cli_runner, tmp_path):
        """
        Tests using the --infer option to add missing values, and also roundtripping
        through yaml->json->yaml->rdf->json
        """
        schema = input_path("schema_with_inference.yaml")
        data_in = input_path("data_example.yaml")
        json_out = tmp_path / "data_example.out.json"
        yaml_out = tmp_path / "data_example.out.yaml"
        rdf_out = tmp_path / "data_example.out.ttl"
        cli_runner.invoke(cli, ["--infer", "-s", schema, data_in, "-o", json_out])
        cli_runner.invoke(cli, ["-s", schema, json_out, "-t", "yaml", "-o", yaml_out])
        cli_runner.invoke(cli, ["-s", schema, yaml_out, "-t", "rdf", "-o", rdf_out])
        cli_runner.invoke(cli, ["-s", schema, rdf_out, "-t", "json", "-o", json_out])
>       with open(json_out) as file:
E       FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest-of-runner/pytest-0/test_infer_and_convert0/data_example.out.json'

```

so the `cli_runner` didn't invoke an error, that was the only test that was sensitive to this change, the change was a change in a docstring, the result was that the file doesn't exist? that's a lot more than a `__str__` method problem

**Comment by Silvanoc** [comment]
I am trying to provide a simple schema to reproduce the issue. That test is not bad to realize that it fails, but error reporting is only showing some symptoms indirectly related to the issue. The CLI call fails due to a JSON-Schema validation (generated JSON-Schema is wrong), therefore the resulting file is missing. Apparently the success of the invocations is not being checked.

I will provide a schema to better illustrate the issue.

**Comment by Silvanoc** [comment]
For easy issue reproduction:

`schema.yaml`
``` yaml
id: https://examples.org/issue2380
name: issue2380

prefixes:
  linkml: https://w3id.org/linkml/
  ex: https://examples.org/issue2280/

default_prefix: ex
default_range: string

imports:
  - linkml:types

classes:
  Person:
    attributes:
      name:
      age_category:
        range: AgeEnum
        ifabsent: AgeEnum(infant)

enums:
  AgeEnum:
    permissible_values:
      infant:
      juvenile:
      adult:
```

`data.yaml`
``` yaml
name: John Doe
```

Running `linkml-convert -f yaml -s schema.yaml -C Person data.yaml` results in the error:
``` shell
[...]
jsonschema.exceptions.ValidationError: {'text': 'infant'} is not one of ['infant', 'juvenile', 'adult']

Failed validating 'enum' in schema['properties']['age_category']:
    {'description': '',
     'enum': ['infant', 'juvenile', 'adult'],
     'title': 'AgeEnum',
     'type': 'string'}

On instance['age_category']:
    {'text': 'infant'}
```

**Comment by sneakers-the-rat** [comment]
i'm running that command directly from commit https://github.com/linkml/linkml/commit/f01fe033344b4c1b2a8759a41848427ea605d1df

what i see is this traceback

```
poetry run linkml convert --infer -s input/schema_with_inference.yaml input/data_example.yaml -o ./data_example.out.json
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-PU7PpyOo-py3.12/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-PU7PpyOo-py3.12/lib/python3.12/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-PU7PpyOo-py3.12/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-PU7PpyOo-py3.12/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-PU7PpyOo-py3.12/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jonny/git/linkml/linkml/utils/converter.py", line 150, in cli
    infer_all_slot_values(obj, schemaview=sv, config=infer_config)
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-PU7PpyOo-py3.12/lib/python3.12/site-packages/linkml_runtime/utils/inference_utils.py", line 150, in infer_all_slot_values
    traverse_object_tree(obj, infer)
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-PU7PpyOo-py3.12/lib/python3.12/site-packages/linkml_runtime/utils/walker_utils.py", line 20, in traverse_object_tree
    return _traverse_object_tree_1(obj, func, mutate)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-PU7PpyOo-py3.12/lib/python3.12/site-packages/linkml_runtime/utils/walker_utils.py", line 36, in _traverse_object_tree_1
    _traverse_object_tree_1(v, func, True)
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-PU7PpyOo-py3.12/lib/python3.12/site-packages/linkml_runtime/utils/walker_utils.py", line 29, in _traverse_object_tree_1
    return {k: _traverse_object_tree_1(v, func, mutate) for k, v in obj.items()}
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-PU7PpyOo-py3.12/lib/python3.12/site-packages/linkml_runtime/utils/walker_utils.py", line 37, in _traverse_object_tree_1
    obj = func(obj)
          ^^^^^^^^^
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-PU7PpyOo-py3.12/lib/python3.12/site-packages/linkml_runtime/utils/inference_utils.py", line 148, in infer
    infer_slot_value(in_obj, k, schemaview, class_name=class_name, policy=policy, config=config)
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-PU7PpyOo-py3.12/lib/python3.12/site-packages/linkml_runtime/utils/inference_utils.py", line 124, in infer_slot_value
    obj.__post_init__()
  File "test", line 172, in __post_init__
TypeError: attribute name must be string, not 'PermissibleValue'
```

aight then so like what is that right thats the next step, i have a debugger, i open it.

blam drop a `pdb.set_trace()` right here: https://github.com/linkml/linkml-runtime/blob/ad74966db72be214a38de19f361df88489286e83/linkml_runtime/utils/inference_utils.py#L124

```python
try:
    obj.__post_init__()
except:
    pdb.set_trace()
```

then i'm all in the debugger like what the heck

```
-> pdb.set_trace()
(Pdb) obj
Person({
  'id': 'P:1',
  'first_name': 'first1',
  'last_name': 'last1',
  'full_name': 'first1 last1',
  'age_in_years': Decimal('10'),
  'age_category': PermissibleValue({'text': 'infant'})
})
(Pdb) type(obj)
<class 'test.Person'>
(Pdb) import inspect
(Pdb) inspect.getsource(obj)
*** TypeError: module, class, method, function, traceback, frame, or code object was expected, got Person
(Pdb) inspect.getsource(type(obj))
*** TypeError: <class 'test.Person'> is a built-in class
```

oh dang, module `test`? `test.Person` appearing as a built-in class to `inspect`? that's gotta be pythongen, classic pythongen signs. so without further ado i hit `u` a bunch of times until i hit the cli scope, and then look at the code and be like ok so the schema is passed in as a path [up here](https://github.com/linkml/linkml/blob/08959cb73d8b09d66bbad730bb5f4d63f0195b08/linkml/utils/converter.py#L108) and i want to see the generated source and so then i just run 

```
print(PythonGenerator(schema).serialize())
```

and i get a whole module. i hit the debugger. things are tense. time is racing. the clock is ticking. I recall that the error message is

```
TypeError: attribute name must be string, not 'PermissibleValue'
```

A hint! I hit `d` a bunch and go back down into the scope where the object is before the `__post_init__` method is called. i look into `obj.__dict__` and there it is!!!

```
(Pdb) obj.__dict__
{'id': 'P:1', 'first_name': 'first1', 'last_name': 'last1', 'full_name': 'first1 last1', 'age_in_years': Decimal('10'), 'age_in_months': None, 'primary_address': None, 'description': None, 'is_juvenile': None, 'age_category': PermissibleValue({'text': 'infant'}), 'summary': None}
```

the bug! `age_category` is being passed in as an instantiated object instead of a string! Wait a second, look back at the generated code, 

```python
        if self.age_category is not None and not isinstance(self.age_category, AgeEnum):
            self.age_category = getattr(AgeEnum, self.age_category)

```

but if `self.age_category` was a `PermissibleValue`, and we call `getattr(AgeEnum, self.age_category)`...

![ToBeContinued LooneyTunes](https://github.com/user-attachments/assets/4978500b-2883-427d-afad-fa9f1cb49e36)

(going to bed)

**Comment by Silvanoc** [comment]
The more I look at it, the weirder I find the notation `AgeEnum(infant)` for `ifabsent` instead of `AgeEnum.infant`. Of course, is it possible to convert the 1st in the 2nd for the Python code generation (and that's what we will need). But it still feels anyway strange. Something like `PermissibleValue(infant)` would have been also fine in my eyes.

**Comment by Silvanoc** [comment]
What is the expectation on the generated Python code?

Following the above example, `type(Person().age_category)` should be `str` or `PermissibleValue`?

Meanwhile it's clear to me that the expectation is that `str(Person().age_category)` should return `infant` (and not `PermissibleValue({'text': 'infant'}`), what is especially relevant for conversion into other formats like JSON.

**Comment by Silvanoc** [comment]
I am sorry, but I've somehow lost my north-star... I do not really know what is the expectation, what is working wrong... I think I need a break from this topic until next week.

In the end I have not been able to fix in v1.8.5 what was broken in v1.8.4, but I have not broken anything new 🙂

**Comment by Silvanoc** [comment]
After having a weekend to let this topic settle in my head, I have now some clarity.

The is my view/understanding of expectations, please correct me if I am wrong. We might need to involve more people to confirm it. It tries to analyze what **looks to me like the root-cause for the issues we are facing trying to fix the `ifabsent` behavior**.

With the schema defined [above in this comment](https://github.com/linkml/linkml/issues/2382#issuecomment-2437523856) and the `person` object of class `Person()`.

# Enumeration

Different input formats result in different objects!! How can it be?! Obviously either the types are not constrained enough, or a necessary conversion is not happening.

TL;DR: my feeling is that the expectation was formats 1 and 3 (possibly even 2 too) to be equivalent, but they have slowly drifted apart.

## Input format 1

`person(..., age_category='infant')`

Aligned with [the documentation](https://linkml.io/linkml/intro/tutorial06.html#use-in-python)

## Input format 2

`person(..., age_category=AgeEnum.infant)`

Behaves exactly like the previous one

# Input format 3

`person(..., age_category=AgeEnum('infant'))`

Behaves differently, as we will see below

# Output

### `repr(person.age_category)`

Format 1 and 2: `"PermissibleValue({'text': 'infant'})"`

Format 3: `AgeEnum(text='infant')`

### `str(person.age_category)`

Format 3: `'infant'`

Format 1 and 2: `"PermissibleValue({'text': 'infant'})"`, differently from the expectation of getting `'infant'`

Fixing it would require a fix in code generation and a regeneration of the meta-model Python code.

In the end, we would only need to modify code generation so that we get [`class PermissibleValue()`](https://github.com/linkml/linkml-runtime/blob/ad74966db72be214a38de19f361df88489286e83/linkml_runtime/linkml_model/meta.py#L4016) having
``` python
  def __str__(self):
    return self.text
```

### `type([...].enum_attr)`

What should yield `type(person.age_category)`?

Thinking on the `range` (an enumeration) is might look as if the result from Format 3 is right: `schema.AgeEnum`. But,

1. it is irritating when compared with the result of `repr(person.age_category)` (which is a permissible value, as you see above)
2. it does not fit what a human being would expect, which is "assign this attribute one of the values among those of this enumeration" and not "assign this attribute an enumeration with an attribute 'code' to specify the desired value".

Formats 1 and 2 result in `class 'PermissibleValue'`, which is a completely different type.

### `type([...].enum_attr.code)`

Since only format 3 provides an `AgeEnum` object, only in that format we have the attribute `code`. The `PermissibleValue` objects that we get in formats 1 and 2 do not have such an attribute.

`type(person.age_category.code)` yields `PermissibleValue({‘text': 'infant'})` so in format 3 you get the same object you get with formats 1 and 2, but as an attribute of an enum.

**Comment by Silvanoc** [comment]
I am basically overwhelmed by the complexity and lack of Python Jedi skills needed for this problem.

**Comment by sneakers-the-rat** [comment]
here i can write a patch, gimme 2 hours :)

**Comment by sneakers-the-rat** [comment]
so interestingly the failure there doesn't seem to have anything to do with pretty printing or the str/repr methods... it should have always failed? as you're saying re: masking, we just didn't have any default enum valued test cases and so no errors were thrown.

The error in the `linkml convert` case above comes from here (which is [curiously not in one of the two packages for validation, `validator` or `validators` but instead in `utils`](https://github.com/linkml/linkml/issues/2371)): https://github.com/linkml/linkml/blob/08959cb73d8b09d66bbad730bb5f4d63f0195b08/linkml/utils/validation.py#L20-L48

Specifically the `_as_dict` call takes the delightfully readable object

```
Person({'name': 'John Doe', 'age_category': PermissibleValue({'text': 'infant'})})
```

into 

```
{'name': 'John Doe', 'age_category': {'text': 'infant'}}
```

as you say, and wandering our way over to the JSON Dumper we find that it just dumps the YamlRoot items as a dictionary like any other: https://github.com/linkml/linkml-runtime/blob/ad74966db72be214a38de19f361df88489286e83/linkml_runtime/dumpers/json_dumper.py#L55

so ya i don't think this would have ever worked, regardless of the changes you or I have made - you just surfaced the problem bc apparently it hasn't been done before?! :)

**Comment by sneakers-the-rat** [comment]
So we've also got a symptom of technical debt happening too, we have a TODO note here to replace the validation framework, so presumably this is an old method and needs to be removed from the library: https://github.com/linkml/linkml/blob/08959cb73d8b09d66bbad730bb5f4d63f0195b08/linkml/utils/converter.py#L154

I have heard this switch mentioned in a number of prior issues, that everything thouls use `validator` and the plugin system, and this seems to have happened quite some time ago: https://github.com/linkml/linkml/issues/911

so I think rather than trying to fix that, we should properly deprecate that module for good.

**Comment by sneakers-the-rat** [comment]
There actually is a permissible value implementation in `linkml-runtime` that's separate from the metamodel one in `linkml_runtime.linkml_model` : https://github.com/linkml/linkml-runtime/blob/main/linkml_runtime/utils/permissiblevalueimpl.py

I am going to suggest in this moment that we take https://github.com/linkml/linkml/issues/2371 seriously and not a pie in the sky idea as pretty necessary in this moment in linkml's history - stuff being lost in a bunch of unsorted modules is really not helping us out here

**Comment by Silvanoc** [comment]
> I am going to suggest in this moment that we take #2371 seriously and not a pie in the sky idea as pretty necessary in this moment in linkml's history - stuff being lost in a bunch of unsorted modules is really not helping us out here

I fully agree. Unfortunately I can only give you one thumb-up and one rocket 😄

The starting steepness of the learning curve is high enough considering that we are writing a meta-model generator for both the meta-model (or schema) desired by the user and the LinkML meta-model being the foundation of the meta-model generator 😵‍💫 

When you start finding modules in unexpected places, duplication and so on an do forth, so start wondering if it's you getting something wrong or if it's something with the code. It takes some time until you understand enough to realize that it's not only you missing something, but also the code with dead parts, duplication and unexpected location.

**Comment by Silvanoc** [comment]
Due to this issue my project is still sticking to `linkml` v1.8.3. @sneakers-the-rat already asked for some input of a _core maintainer_ in [this comment](https://github.com/linkml/linkml/issues/2382#issuecomment-2442434381) to move on. But, who is a "core maintainer"? Are they listed somewhere? Apart from @cmungall I could not name any...

I think I already mentioned it before somewhere else, but I'll repeat it. IMO the project need to explicitly define roles, people playing those roles and how these roles operate on the different decision taking situations both at operative level (PRs) and at strategic level.

**Comment by dalito** [comment]
The current maintainers are listed here: https://github.com/orgs/linkml/teams/maintainers

...but it does not tell whom to ask for what. This could maybe be done with [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) file.

**Comment by Silvanoc** [comment]
> The current maintainers are listed here: https://github.com/orgs/linkml/teams/maintainers

Those are the users with maintainer rights in the GitHub organization. BTW, I am not sure that list is being really maintained (kept up to date), since some of those people seem to be absent from the project for a long time.

But @dalito and @sneakers-the-rat are using the term "core maintainers" (last time [here](https://github.com/linkml/linkml-runtime/pull/364#issuecomment-2623925810)) as if they were a subset of that group. I have also the feeling that is is only a subset of that group feeling confident to take design decisions with big impact in the project.

> ...but it does not tell whom to ask for what. This could maybe be done with [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) file.

I would also like to see that kind of information documented at code level in the affected repositories.

I though I had already opened a discussion for this topic. Apparently I hadn't yet, only on Slack. So now I have created a GitHub Discussion for it 🙂 : https://github.com/orgs/linkml/discussions/2518

**Comment by amc-corey-cox** [comment]
Thanks for the thorough investigation here, and apologies for the long stall — this has been waiting on maintainer input and that's on us.

I think there are two viable paths forward:

**A) Fix in [`remove_empty_items`](https://github.com/linkml/linkml/blob/ccc7546657a73ac435d969a16c1f6aca6e276de6/packages/linkml_runtime/src/linkml_runtime/utils/formatutils.py#L124)** — there's already a [special case for enum serialization that looks for a `_code` key](https://github.com/linkml/linkml/blob/ccc7546657a73ac435d969a16c1f6aca6e276de6/packages/linkml_runtime/src/linkml_runtime/utils/formatutils.py#L178), but `PermissibleValue` serializes as `{'text': 'active'}` so it doesn't match. Updating that detection logic would fix all serialization paths at once. Pragmatic, but adds more magic to an already-magic function.

**B) Fix in [`EnumDefinitionImpl`](https://github.com/linkml/linkml/blob/ccc7546657a73ac435d969a16c1f6aca6e276de6/packages/linkml_runtime/src/linkml_runtime/utils/enumerations.py#L37)** — teach the enum base class how its members should serialize, so the dumpers do the right thing without pattern-matching on dict shape. Cleaner, but might be harder depending on `jsonasobj2` internals.

@sneakers-the-rat @Silvanoc — do either of you have a preference, or is there a better path I haven't thought of? Happy to take this to a dev call if you'd like more eyes on it, but also happy to approve whichever approach you both agree on.

**Comment by Silvanoc** [comment]
> **B) Fix in [`EnumDefinitionImpl`](https://github.com/linkml/linkml/blob/ccc7546657a73ac435d969a16c1f6aca6e276de6/packages/linkml_runtime/src/linkml_runtime/utils/enumerations.py#L37)** — teach the enum base class how its members should serialize, so the dumpers do the right thing without pattern-matching on dict shape. Cleaner, but might be harder depending on `jsonasobj2` internals.

I would go for this option, because it looks to me like the right way to fix it and the code has already enough magic functions that put the entry bar pretty high for new contributors.

If `jsonasobj2` is a blocker here, I'd say it's just another nail in `jsonasobj2`'s coffin in LinkML... I think that we have enough reasons to get rid of it (see [here1](https://github.com/linkml/linkml/issues/2683#issuecomment-2900390625), [here2](https://github.com/linkml/linkml/issues/2217) and [all the related issues](https://github.com/linkml/linkml/issues?q=is%3Aissue%20state%3Aopen%20jsonasobj2)).

@amc-corey-cox I've edited your previous comment to put links to the code for more clarity, hope that you are fine with it and that the links a correct.

**Comment by Silvanoc** [comment]
> Thanks for the thorough investigation here, and apologies for the long stall — this has been waiting on maintainer input and that's on us.

No problem, there's so much to do! I know that you are not idling.

**Comment by sneakers-the-rat** [comment]
I guess my preference would be to just use regular enums, but aside from that, i also would go with just fixing the str representation in the enum definition implementation class, though i admit it's been too long since i thought about this to have much of an informed opinion
